File: Algorithms/402. Remove K Digits/golang/solution.go
func removeKdigits(num string, k int) string {
	numStack := []rune{}

	for _, digit := range num {
		for k > 0 && len(numStack) > 0 && numStack[len(numStack)-1] > digit {
			numStack = numStack[:len(numStack)-1]
			k--
		}

		numStack = append(numStack, digit)
	}

	for k > 0 {
		numStack = numStack[:len(numStack)-1]
		k--
	}

	for len(numStack) > 0 && numStack[0] == '0' {
		numStack = numStack[1:len(numStack)]
	}

	if len(numStack) == 0 {
		return "0"
	}
	return string(numStack)
}

File: Algorithms/402. Remove K Digits/javascript/solution.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function (num, k) {
    const stack = [];

    for (const digit of num) {
        while (k && stack && stack[stack.length - 1] > digit) {
            stack.pop();
            k--;
        }
        stack.push(digit);
    }

    const finalStack = k ? stack.slice(0, -k) : stack;
    return finalStack.join('').replace(/^0+/, '') || '0';
};
